src/Components/Server/src/Circuits/RemoteJSRuntime.cs

```diff
@@ -78,17 +79,32 @@ protected override void BeginInvokeJS(long asyncHandle, string identifier, strin
         {
             if (_clientProxy is null)
             {
-                throw new InvalidOperationException(
-                    "JavaScript interop calls cannot be issued at this time. This is because the component is being " +
-                    $"statically rendered. When prerendering is enabled, JavaScript interop calls can only be performed " +
-                    $"during the OnAfterRenderAsync lifecycle method.");
+                if (_permanentlyDisconnected)
+                {
+                    throw new JSDisconnectedException(
+                   "JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected " +
+                   "and is being disposed.");
+                }
+                else
+                {
+                    throw new InvalidOperationException(
+                        "JavaScript interop calls cannot be issued at this time. This is because the component is being " +
+                        $"statically rendered. When prerendering is enabled, JavaScript interop calls can only be performed " +
+                        $"during the OnAfterRenderAsync lifecycle method.");
+                }
             }
 
             Log.BeginInvokeJS(_logger, asyncHandle, identifier);
 
             _clientProxy.SendAsync("JS.BeginInvokeJS", asyncHandle, identifier, argsJson, (int)resultType, targetInstanceId);
         }
 
+        public void MarkPermanentlyDisconnected()
+        {
+            _permanentlyDisconnected = true;
```

Comment:
We aren't concerned about a potential race condition between this func and https://github.com/dotnet/aspnetcore/pull/32901/files#diff-7fac6aa482260bd9b68de9fd59db0da9d8532c59ecdfd4468c0f51d371b562aaR82 because we assume a single thread, correct?

Reply:
Great question, and yes, we absolutely expect code invoking the JS runtime to be on the circuit sync context, so only one at a time.